src/szk.rs

```diff
@@ -0,0 +1,574 @@
+        xof.into_seed()
+    }
+
+    fn derive_joint_rand(
+        &self,
+        leader_joint_rand_part: &Seed<SEED_SIZE>,
+        helper_joint_rand_part: &Seed<SEED_SIZE>,
+    ) -> (Seed<SEED_SIZE>, Vec<T::Field>) {
+        let joint_rand_seed =
+            self.derive_joint_rand_seed(leader_joint_rand_part, helper_joint_rand_part);
+        let joint_rand = P::seed_stream(
+            &joint_rand_seed,
+            &self.domain_separation_tag(DST_JOINT_RANDOMNESS),
+            &[],
+        )
+        .into_field_vec(self.typ.joint_rand_len());
+
+        (joint_rand_seed, joint_rand)
+    }
+
+    fn derive_helper_proof_share(&self, proof_share_seed: &Seed<SEED_SIZE>) -> Vec<T::Field> {
+        Prng::from_seed_stream(P::seed_stream(
+            proof_share_seed,
+            &self.domain_separation_tag(DST_PROOF_SHARE),
+            &[],
+        ))
+        .take(self.typ.proof_len())
+        .collect()
+    }
+
+    fn derive_query_rand(&self, verify_key: &[u8; SEED_SIZE], nonce: &[u8; 16]) -> Vec<T::Field> {
+        let mut xof = P::init(
+            verify_key,
+            &self.domain_separation_tag(DST_QUERY_RANDOMNESS),
+        );
+        xof.update(nonce);
+        xof.into_seed_stream()
+            .into_field_vec(self.typ.query_rand_len())
+    }
+
+    fn prove(
+        &self,
+        leader_input_share: &[T::Field],
+        helper_input_share: &[T::Field],
+        encoded_measurement: &[T::Field],
+        rand_seeds: [Seed<SEED_SIZE>; 2],
+        leader_seed_opt: Option<Seed<SEED_SIZE>>,
+        nonce: &[u8; 16],
+    ) -> Result<[SzkProofShare<T::Field, SEED_SIZE>; 2], SzkError> {
+        let [prove_rand_seed, helper_seed] = rand_seeds;
+        let leader_seed_opt_copy = leader_seed_opt.clone();
+        // If joint randomness is used, derive it from the two input shares,
+        // the seeds used to blind the derivation, and the nonce. Pass the
+        // leader its blinding seed and the helper's joint randomness part, and
+        // pass the helper the leader's joint randomness part. (The seed used to
+        // derive the helper's proof share is reused as the helper's blind)
+        let (leader_blind_and_helper_joint_rand_part, leader_joint_rand_part, joint_rand) =
+            if let Some(leader_seed) = leader_seed_opt_copy {
+                let leader_joint_rand_part =
+                    self.derive_joint_rand_part(&leader_seed, leader_input_share, nonce)?;
+                let helper_joint_rand_part =
+                    self.derive_joint_rand_part(&helper_seed, helper_input_share, nonce)?;
+                let (_jrs, jr) =
+                    self.derive_joint_rand(&leader_joint_rand_part, &helper_joint_rand_part);
+                (
+                    Some((leader_seed, helper_joint_rand_part)),
+                    Some(leader_joint_rand_part),
+                    jr,
+                )
+            } else {
+                (None, None, Vec::<T::Field>::new())
```

Comment:
style: Let the compiler infer the type here.
```suggestion
                (None, None, Vec::new())
```